File: RmlUi.Native/src/Rml.cpp
#include "RmlNative.h"
#include "RmlUi/Core.h"
#include "EventListener.h"

RMLUI_CAPI void rml_Initialise() {
    Rml::Initialise();
}

RMLUI_CAPI void rml_Shutdown() {
    Rml::Shutdown();
}

RMLUI_CAPI void rml_SetRenderInterface(Rml::RenderInterface *renderInterface) {
    Rml::SetRenderInterface(renderInterface);
}

RMLUI_CAPI void rml_SetSystemInterface(Rml::SystemInterface *systemInterface) {
    Rml::SetSystemInterface(systemInterface);
}

RMLUI_CAPI void rml_LoadFontFace(const char *fileName, bool fallbackFace, Rml::Style::FontWeight weight) {
    Rml::LoadFontFace(fileName, fallbackFace, weight);
}

RMLUI_CAPI Rml::EventListener *rml_CreateEventListener(::onProcessEvent onProcessEvent, ::onAttachEvent onAttachEvent, ::onDetachEvent onDetachEvent) {
    return new EventListenerProxy(onProcessEvent, onAttachEvent, onDetachEvent);
}

RMLUI_CAPI void rml_ReleaseEventListener(Rml::EventListener *eventListener) {
    delete eventListener;
}

RMLUI_CAPI void rml_RemoveContext(const char *name) {
    Rml::RemoveContext(name);
}

File: RmlUi.Net/Rml.cs
namespace ChickenWithLips.RmlUi;

public static class Rml
{
    public static void Initialise()
    {
        Native.Rml.Initialise();
    }

    public static void Shutdown()
    {
        Native.Rml.Shutdown();
    }

    public static void SetSystemInterface(SystemInterface systemInterface)
    {
        Native.Rml.SetSystemInterface(systemInterface.NativePtr);
    }

    public static void SetRenderInterface(RenderInterface renderInterface)
    {
        Native.Rml.SetRenderInterface(renderInterface.NativePtr);
    }

    /// <summary>
    /// Adds a new font face to the font engine. The face's family, style and weight will be determined from the face itself.
    /// </summary>
    /// <param name="fileName">The file to load the face from.</param>
    /// <param name="fallbackFace">True to use this font face for unknown characters in other font faces.</param>
    /// <param name="weight">The weight to load when the font face contains multiple weights, otherwise the weight to register the font as. By default it loads all found font weights.</param>
    /// <returns>True if the face was loaded successfully, false otherwise.</returns>
    public static bool LoadFontFace(string fileName, bool fallbackFace = false, FontWeight weight = FontWeight.Auto)
    {
        return Native.Rml.LoadFontFace(fileName, fallbackFace, weight);
    }

    public static Context? CreateContext(string name, Vector2i dimensions, RenderInterface? renderInterface = null)
    {
        return Context.Create(name, dimensions, renderInterface);
    }

    public static bool RemoveContext(string name)
    {
        return Native.Rml.RemoveContext(name);
    }
}
